File: android/keyple-plugin/android-omapi/src/test/java/org/eclipse/keyple/plugin/android/omapi/AndroidOmapiReaderTest.java
/********************************************************************************
 * Copyright (c) 2018 Calypso Networks Association https://www.calypsonet-asso.org/
 *
 * See the NOTICE file(s) distributed with this work for additional information regarding copyright
 * ownership.
 *
 * This program and the accompanying materials are made available under the terms of the Eclipse
 * Public License 2.0 which is available at http://www.eclipse.org/legal/epl-2.0
 *
 * SPDX-License-Identifier: EPL-2.0
 ********************************************************************************/
package org.eclipse.keyple.plugin.android.omapi;


import static org.powermock.api.mockito.PowerMockito.when;
import java.io.IOException;
import java.util.*;

import org.eclipse.keyple.core.seproxy.SeSelector;
import org.eclipse.keyple.core.seproxy.exception.KeypleBaseException;
import org.eclipse.keyple.core.seproxy.exception.KeypleIOReaderException;
import org.eclipse.keyple.core.seproxy.exception.KeypleReaderException;
import org.eclipse.keyple.core.seproxy.message.ApduRequest;
import org.eclipse.keyple.core.seproxy.message.SeRequest;
import org.eclipse.keyple.core.seproxy.message.SeResponse;
import org.eclipse.keyple.core.seproxy.protocol.SeCommonProtocols;
import org.eclipse.keyple.core.util.ByteArrayUtil;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mockito;
import org.powermock.modules.junit4.PowerMockRunner;
import org.simalliance.openmobileapi.Channel;
import org.simalliance.openmobileapi.Reader;
import org.simalliance.openmobileapi.Session;

@RunWith(PowerMockRunner.class)
public class AndroidOmapiReaderTest {

    final String PLUGIN_NAME = "AndroidOmapiPluginImpl";
    final String poAid = "A000000291A000000191";
    final String poAidResponse =
            "6F25840BA000000291A00000019102A516BF0C13C70800000000C0E11FA653070A3C230C1410019000";

    Reader omapiReader;
    AndroidOmapiReaderImpl reader;

    // init before each test
    @Before
    public void SetUp() throws Exception {
        // default reader connected with secure element with poAid
        omapiReader = mockReader();

        // instantiate reader with omapiReader
        reader = new AndroidOmapiReaderImpl(PLUGIN_NAME, omapiReader, omapiReader.getName());
    }


    /*
     * TEST READERS
     */

    @Test
    public void getInstance() throws Exception {
        Assert.assertNotNull(reader);
    }

    @Test
    public void getName() throws Exception {
        Assert.assertEquals(omapiReader.getName(), reader.getName());
    }

    @Test
    public void isSEPresent() throws KeypleIOReaderException {
        Assert.assertEquals(true, reader.isSePresent());
    }

    @Test
    public void getParameters() {
        Assert.assertNotNull(reader.getParameters());
    }

    @Test
    public void setParameters() throws KeypleBaseException {
        Map<String, String> parameters = new HashMap<String, String>();
        parameters.put("key1", "value1");
        reader.setParameters(parameters);
        Assert.assertTrue(reader.getParameters().size() == 1);
        Assert.assertTrue(reader.getParameters().get("key1").equals("value1"));
    }

    @Test
    public void setParameter() throws KeypleBaseException {
        reader.setParameter("key2", "value2");
        Assert.assertTrue(reader.getParameters().size() == 1);
        Assert.assertTrue(reader.getParameters().get("key2").equals("value2"));
    }

    /*
     * TRANSMIT
     */


    @Test
    public void transmitNoAid() throws KeypleBaseException, IOException {

        // init
        omapiReader = mockReaderWithNoAid();
        reader = new AndroidOmapiReaderImpl(PLUGIN_NAME, omapiReader, omapiReader.getName());

        // test
        List<SeResponse> seResponseList = reader.transmitSet(getSampleSeRequest());

        // assert
        Assert.assertNull(seResponseList.get(0));

    }

    @Test
    public void transmitWrongProtocol() throws KeypleBaseException {
        // init
        String poAid = "A000000291A000000191";

        // wrong protocol
        SeRequest seRequest = new SeRequest(new SeSelector( SeCommonProtocols.PROTOCOL_MIFARE_UL, null,
                new SeSelector.AidSelector(new SeSelector.AidSelector.IsoAid(poAid),null),null), new ArrayList<ApduRequest>());

        // test
        Set<SeRequest> seRequestSet = new LinkedHashSet<SeRequest>();
        seRequestSet.add(seRequest);
        List<SeResponse> seResponseList = reader.transmitSet(seRequestSet);

        // assert
        Assert.assertNull(seResponseList.get(0));

    }

    @Test(expected = KeypleReaderException.class)
    public void transmitNotConnected() throws KeypleBaseException, IOException {

        // init
        Reader omapiReader = Mockito.mock(Reader.class);
        when(omapiReader.getName()).thenReturn("SIM1");
        when(omapiReader.isSecureElementPresent()).thenReturn(false);
        when(omapiReader.openSession()).thenThrow(new IOException());
        reader = new AndroidOmapiReaderImpl(PLUGIN_NAME, omapiReader, omapiReader.getName());

        // test
        List<SeResponse> seResponseList = reader.transmitSet(getSampleSeRequest());

        // expected = KeypleReaderException.class
    }



    /*
     * HELPERS
     */


    Reader mockReader() throws IOException {

        Reader omapiReader = Mockito.mock(Reader.class);
        Session session = Mockito.mock(Session.class);
        Channel channel = Mockito.mock(Channel.class);

        when(omapiReader.getName()).thenReturn("SIM1");
        when(omapiReader.isSecureElementPresent()).thenReturn(true);
        when(session.openLogicalChannel(ByteArrayUtil.fromHex(poAid), (byte)0x00)).thenReturn(channel);
        when(omapiReader.openSession()).thenReturn(session);
        when(session.getATR()).thenReturn(null);
        when(channel.getSelectResponse()).thenReturn(ByteArrayUtil.fromHex(poAidResponse));
        when(channel.getSession()).thenReturn(session);

        when(channel.transmit(ByteArrayUtil.fromHex("00B201A420"))).thenReturn(ByteArrayUtil
                .fromHex("00000000000000000000000000000000000000000000000000000000000000009000"));

        return omapiReader;

    }

    Reader mockReaderWithNoAid() throws KeypleBaseException, IOException {

        Reader omapiReader = Mockito.mock(Reader.class);
        Session session = Mockito.mock(Session.class);
        Channel channel = Mockito.mock(Channel.class);

        when(omapiReader.getName()).thenReturn("SIM1");
        when(omapiReader.isSecureElementPresent()).thenReturn(true);
        when(omapiReader.openSession()).thenReturn(session);
        when(session.openLogicalChannel(ByteArrayUtil.fromHex(poAid), (byte)0x00))
                .thenThrow(new NoSuchElementException(""));

        return omapiReader;

    }

    Set<SeRequest> getSampleSeRequest() {
        String poAid = "A000000291A000000191";

        List<ApduRequest> poApduRequestList  = Arrays.asList(new ApduRequest(ByteArrayUtil.fromHex("0000"), true));

        SeRequest seRequest = new SeRequest(new SeSelector(SeCommonProtocols.PROTOCOL_ISO7816_3, null,
                new SeSelector.AidSelector(new SeSelector.AidSelector.IsoAid(poAid), null), null), poApduRequestList);

        Set<SeRequest> seRequestSet = new LinkedHashSet<SeRequest>();
        seRequestSet.add(seRequest);
        return seRequestSet;

    }


}

File: android/keyple-plugin/android-omapi/src/test/kotlin/org/eclipse/keyple/plugin/android/omapi/AndroidOmapiReaderTest.kt
package org.eclipse.keyple.plugin.android.omapi

import io.mockk.MockKAnnotations
import io.mockk.unmockkAll
import java.io.IOException
import kotlin.NoSuchElementException
import org.eclipse.keyple.core.seproxy.ChannelControl
import org.eclipse.keyple.core.seproxy.MultiSeRequestProcessing
import org.eclipse.keyple.core.seproxy.SeSelector
import org.eclipse.keyple.core.seproxy.exception.KeypleChannelControlException
import org.eclipse.keyple.core.seproxy.exception.KeypleIOReaderException
import org.eclipse.keyple.core.seproxy.exception.KeypleReaderException
import org.eclipse.keyple.core.seproxy.message.ApduRequest
import org.eclipse.keyple.core.seproxy.message.SeRequest
import org.eclipse.keyple.core.seproxy.protocol.SeCommonProtocols
import org.eclipse.keyple.core.seproxy.protocol.TransmissionMode
import org.eclipse.keyple.core.util.ByteArrayUtil
import org.junit.After
import org.junit.Assert
import org.junit.Before
import org.junit.Ignore
import org.junit.Test

internal abstract class AndroidOmapiReaderTest<T, V : AndroidOmapiReader> {

    companion object {
        internal const val PLUGIN_NAME = "AndroidOmapiPluginImpl"
        internal const val PO_AID = "A000000291A000000191"
        internal const val PO_AID_RESPONSE = "6F25840BA000000291A00000019102A516BF0C13C70800000000C0E11FA653070A3C230C1410019000"
    }

    abstract var nativeReader: T
    abstract var reader: V

    abstract fun mockReader(): T
    abstract fun mockReaderWithExceptionOnOpenLogicalChannel(throwable: Throwable): T
    abstract fun mockReaderWithNullOnOpenLogicalChannel(): T
    abstract fun mockReaderWithExceptionOnOpenBasicChannel(throwable: Throwable): T
    abstract fun mockReaderWithNullOnOpenBasicChannel(): T
    abstract fun mockReaderWithNoAid(): T
    abstract fun mockReaderWithExceptionOnOpenSession(throwable: Throwable): T
    abstract fun buildOmapiReaderImpl(nativeReader: T): V
    abstract fun getNativeReaderName(): String
    abstract fun mockReaderWithExceptionOnCloseChannel(throwable: Throwable): T
    abstract fun mockReaderWithExceptionWhileTransmittingApdu(throwable: Throwable): T

    @Before
    fun setUp() {
        MockKAnnotations.init(this, relaxUnitFun = true)
        // default reader connected with secure element with poAid
        nativeReader = mockReader()

        // instantiate reader with nativeReader
        reader = buildOmapiReaderImpl(nativeReader)
    }

    @After
    fun tearDown() {
        unmockkAll()
    }
    /*
     * TEST READERS
     */

    @Test
    fun getInstance() {
        Assert.assertNotNull(reader)
    }

    @Test
    fun getName() {
        Assert.assertEquals(getNativeReaderName(), reader.name)
    }

    @Test
    fun getTransmissionMode() {
        Assert.assertEquals(TransmissionMode.CONTACTS, reader.transmissionMode)
    }

    @Test
    fun isSEPresent() {
        Assert.assertEquals(true, reader.isSePresent)
    }

    @Test
    fun getParameters() {
        Assert.assertNotNull(reader.parameters)
    }

    @Test
    fun setParameters() {
        val parameters = HashMap<String, String>()
        parameters["key1"] = "value1"
        reader.parameters = parameters
        Assert.assertTrue(reader.parameters.size == 1)
        Assert.assertTrue(reader.parameters["key1"] == "value1")
    }

    @Test
    fun setParameter() {
        reader.setParameter("key2", "value2")
        Assert.assertTrue(reader.parameters.size == 1)
        Assert.assertTrue(reader.parameters["key2"] == "value2")
    }

    @Test(expected = KeypleIOReaderException::class)
    fun transmitIOException() {
        nativeReader = mockReaderWithExceptionOnOpenLogicalChannel(IOException())
        reader = buildOmapiReaderImpl(nativeReader)
        // test
        reader.transmitSet(getSampleSeRequest())
    }

    @Test(expected = KeypleChannelControlException::class)
    fun transmitSecurityException() {
        nativeReader = mockReaderWithExceptionOnOpenLogicalChannel(SecurityException())
        reader = buildOmapiReaderImpl(nativeReader)
        // test
        reader.transmitSet(getSampleSeRequest())
    }

    @Test
    fun transmitNoSuchElementException() {
        nativeReader = mockReaderWithExceptionOnOpenLogicalChannel(NoSuchElementException())
        reader = buildOmapiReaderImpl(nativeReader)
        // test
        val seResponseList = reader.transmitSet(getSampleSeRequest())
        Assert.assertNull(seResponseList[0]) // If container is not found a null responsed is returned
    }

    @Test(expected = KeypleIOReaderException::class)
    fun transmitOpenChannelFailed() {
        nativeReader = mockReaderWithNullOnOpenLogicalChannel()
        reader = buildOmapiReaderImpl(nativeReader)
        // test
        reader.transmitSet(getSampleSeRequest())
    }

    @Test(expected = KeypleIOReaderException::class)
    fun transmitNoAidOpenChannelFailed() {
        nativeReader = mockReaderWithNullOnOpenBasicChannel()
        reader = buildOmapiReaderImpl(nativeReader)
        // test
        reader.transmitSet(getNoAidSampleSeRequest())
    }

    @Test(expected = KeypleIOReaderException::class)
    fun transmitNoAidIOException() {
        nativeReader = mockReaderWithExceptionOnOpenBasicChannel(IOException())
        reader = buildOmapiReaderImpl(nativeReader)
        // test
        reader.transmitSet(getNoAidSampleSeRequest())
    }

    @Test(expected = KeypleChannelControlException::class)
    fun transmitNoAidSecurityException() {
        nativeReader = mockReaderWithExceptionOnOpenBasicChannel(SecurityException())
        reader = buildOmapiReaderImpl(nativeReader)
        // test
        reader.transmitSet(getNoAidSampleSeRequest())
    }

    @Test
    fun transmitNoAid() {

        // init
        nativeReader = mockReaderWithNoAid()
        reader = buildOmapiReaderImpl(nativeReader)

        // test
        val seResponseList = reader.transmitSet(getSampleSeRequest())

        // assert
        Assert.assertNull(seResponseList[0])
    }

    @Test
    fun transmitWrongProtocol() {
        // init
        val poAid = "A000000291A000000191"

        // wrong protocol
        val seRequest = SeRequest(SeSelector(SeCommonProtocols.PROTOCOL_MIFARE_UL, null,
                SeSelector.AidSelector(SeSelector.AidSelector.IsoAid(poAid), null), null), ArrayList())

        // test
        val seRequestSet = LinkedHashSet<SeRequest>()
        seRequestSet.add(seRequest)
        val seResponseList = reader.transmitSet(seRequestSet)

        // assert
        Assert.assertNull(seResponseList[0])
    }

    @Test(expected = KeypleReaderException::class)
    fun transmitNotConnected() {

        // init
        val nativeReader = mockReaderWithExceptionOnOpenSession(IOException())
        reader = buildOmapiReaderImpl(nativeReader)

        // test
        reader.transmitSet(getSampleSeRequest())
    }

    @Test
    fun closeChannelAfterTransmit() {

        // init
        val nativeReader = mockReader()
        reader = buildOmapiReaderImpl(nativeReader)

        // test
        val seResponseList = reader.transmitSet(getSampleSeRequest(), MultiSeRequestProcessing.PROCESS_ALL, ChannelControl.CLOSE_AFTER)
        Assert.assertNotNull(seResponseList)
    }

    @Test(expected = KeypleIOReaderException::class)
    @Ignore("Implementation of closePhysicalChannel() may encounter a IOException, it should be catched as a KeypleIOReaderException, it is not")
    fun exceptionOnChannelAfterTransmit() {
        // init
        val nativeReader = mockReaderWithExceptionOnCloseChannel(IOException())
        reader = buildOmapiReaderImpl(nativeReader)

        // test
        reader.transmitSet(getSampleSeRequest(), MultiSeRequestProcessing.PROCESS_ALL, ChannelControl.CLOSE_AFTER)
    }

    @Test(expected = KeypleIOReaderException::class)
    fun exceptionWhileTransmittingApdu() {
        // init
        val nativeReader = mockReaderWithExceptionWhileTransmittingApdu(IOException())
        reader = buildOmapiReaderImpl(nativeReader)

        // test
        reader.transmitSet(getSampleSeRequest())
    }

    private fun getSampleSeRequest(): Set<SeRequest> {

        val poApduRequestList = listOf(ApduRequest(ByteArrayUtil.fromHex("0000"), true))

        val seRequest = SeRequest(SeSelector(SeCommonProtocols.PROTOCOL_ISO7816_3, null,
                SeSelector.AidSelector(SeSelector.AidSelector.IsoAid(PO_AID), null), null), poApduRequestList)

        val seRequestSet = LinkedHashSet<SeRequest>()
        seRequestSet.add(seRequest)
        return seRequestSet
    }

    private fun getNoAidSampleSeRequest(): Set<SeRequest> {

        val poApduRequestList = listOf(ApduRequest(ByteArrayUtil.fromHex("0000"), true))

        val seRequest = SeRequest(SeSelector(SeCommonProtocols.PROTOCOL_ISO7816_3, null,
                SeSelector.AidSelector(null, null), null), poApduRequestList)

        val seRequestSet = LinkedHashSet<SeRequest>()
        seRequestSet.add(seRequest)
        return seRequestSet
    }
}
